Netezos/IO/MichelineReader.cs

```diff
@@ -0,0 +1,447 @@
+﻿using Netezos.Encoding;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Numerics;
+
+namespace Netezos.IO
+{
```

Comment:
I decided to go with a `MichelineReader` type that just acts like a stream reader for primitive types (rather than trying to read byte spans or anything like that). It was an easy way to track position and consume bytes as needed. It works well for nested types too, as you can see when reading arrays. All data from the array can be read by reading bytes equal to the length of the array, wrapping those bytes in another reader, and reading the types until the end of the stream.